### ISSUE TITLE
Add support for x-pack (security)

### ISSUE THREAD

**Issue by dadoonet** [alexcojocaru/elasticsearch-maven-plugin#42]
When you add x-pack as a plugin, elasticsearch maven plugin refuses to start because it can't access to `GET /` due to default security.

It gives:

```
[DEBUG] Sending GET request to http://localhost:9400/
[DEBUG] Response with status code 401 and content: {"error":{"root_cause":[{"type":"security_exception","reason":"missing authentication token for REST request [/]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}}],"type":"security_exception","reason":"missing authentication token for REST request [/]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}},"status":401}
```

We need to add support for username and password (default to `elastic` and `changeme`).

I'll hopefully will come with a PR for it.

**Comment by lukas-vlcek** [comment]
@dadoonet I do not want to hijack this thread but by any chance would you be able to direct me to example how this could done with Elastic's own integration testing framework? In other words how to get a `client()` in [`ESIntegTestCase`](https://www.elastic.co/guide/en/elasticsearch/reference/5.0/integration-tests.html) and make sure this client will be using some kind of authentication? TLS for example? I did not find any documentation around this.

BTW, is there any reason why you use this plugin in your [IT tests](https://github.com/dadoonet/elasticsearch-beyonder/tree/master/src/test/java/fr/pilato/elasticsearch/tools) and do not use mentioned `ESIntegTestCase`?

**Comment by dadoonet** [comment]
Hi @lukas-vlcek. I'm not using myself on external projects the `ESIntegTestCase` and globally the elasticsearch test framework.
Not sure if it is meant to be used outside the context of elasticsearch project TBH.

This is why I'm not using this class. I found much easier to start my elasticsearch instance externally (I have been doing that using my own ANT script and now I switched to this project which does all that in a much better way).
I might give a look at https://github.com/allegro/embedded-elasticsearch at some point though. Not sure yet.

**Comment by lukas-vlcek** [comment]
@dadoonet thanks. In my case I am trying to implement ITs for ES custom plugin but it also involves security context provided by other plugin (SG), so that is why I am curious how one can handle that in ESIntegTestCase approach.

**Comment by dadoonet** [comment]
Something is changing with elasticsearch 6.0.0-beta1. There is no more default password `changeme`.

If people want to use x-pack, they need to define a password as follows:

```sh
bin/elasticsearch-keystore create
echo "changeme" | bin/elasticsearch-keystore add "bootstrap.password" --stdin
```

Also it might be needed to start elasticsearch with the following option:

```yml
xpack.security.authc.accept_default_password: true
```

Or 

```sh
bin/elasticsearch -Expack.security.authc.accept_default_password=true
```

This complicates a bit the maven plugin. I wonder if we should add something specific for x-pack which will do that OOTB (my preference) or something more generic like a `post_install_script` option and a global `esJavaOpts` option which can be appended when launching elasticsearch.

Or both of them... Thoughts?

**Comment by alexcojocaru** [comment]
If I read your comment correctly: what you are saying is that, with the above 2 commands, x-pack can be installed.

When does the first set of commands (ie. creating the keystore and adding the password to it) need to be executed? Before starting the ES instance?
Can the 2nd command be rewritten not to use the pipe? The cli framework is very sensitive and it is possible it won't work in that format. And it should work on Windows too.

The argument to set the `accept_default_password` to `true` could be set always on the command - do you see any reasons for not doing so?

Also, there is the problem with having the plugin work with ES6 and also keeping it compatible with ES5 - I thought that, by rewritting the plugin, I could get support for major version changes for free, but apparently the ES devs are doing their best at breaking backwards compatibility.

**Comment by dadoonet** [comment]
> When does the first set of commands (ie. creating the keystore and adding the password to it) need to be executed? Before starting the ES instance?

Yes. And it can take some time before the user is actually created. So it's needed to test and retry before failing.

> Can the 2nd command be rewritten not to use the pipe?

I'm afraid it's not possible. Without the --stdin it will be asked to the user to enter the password (the setting value). Not sure it will work better.

> do you see any reasons for not doing so?

Yes. If x-pack is not installed, elasticsearch will refuse to start as the setting will be unknown.

> but apparently the ES devs are doing their best at breaking backwards compatibility.

I'm sad you feel it this way. That's not true IMHO. 
What you are doing with your great plugin is not a "common" API usage of elasticsearch. That's why it can break sometime.
Also, for the default password for example, we thought it was not safe to keep it around.
I'm pretty sure it will become more and more mature with time.

**Comment by dadoonet** [comment]
> Actually, we don't need to set `accept_default_password` to `true`. Which means that we only have to set `bootstrap.password` which can always be set in 6.0.0 even if you don't install x-pack.
(Still doing some tests around this).

Argh! Scratch this. The `bootstrap.password` is tight to x-pack. If you set it but don't run elasticsearch with x-pack, it will fail with:

```
[elasticsearch] [2017-08-11T16:26:33,252][WARN ][o.e.b.ElasticsearchUncaughtExceptionHandler] [] uncaught exception in thread [main]
[elasticsearch] org.elasticsearch.bootstrap.StartupException: java.lang.IllegalArgumentException: unknown secure setting [bootstrap.password] please check that any required plugins are installed, or check the breaking changes documentation for removed settings
```

**Comment by alexcojocaru** [comment]
This issue is close to 5 years old. A lot has changes in the ES world since then; I am closing it.
